contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,628 @@
+     */
+    function rebalance(uint256 _amountToSwap, uint256 _minTokenReceived, bool _swapWETH) external nonReentrant onlyGovernorOrStrategist {
+        _rebalance(_amountToSwap, _minTokenReceived, _swapWETH);
+    }
+
+    /**
+     * @dev Rebalance the pool to the desired token split
+     */
+    function _rebalance(uint256 _amountToSwap, uint256 _minTokenReceived, bool _swapWETH) internal {
+        _removeLiquidity();
+        _swapToDesiredPosition(_amountToSwap, _minTokenReceived, _swapWETH);
+        _addLiquidity();
+        _checkLiquidityWithinExpectedShare();
+    }
+
+    /**
+     * @dev Decrease withdrawLiquidityShare (currently set to 99%) of the liquidity if strategy holds any. In practice the removal of liquidity
+     * will be skipped only on the first time called.
+     */
+    function _removeLiquidity() internal {
+        if (tokenId == 0) {
+            return;
+        }
+
+        clGauge.withdraw(tokenId);
+
+        (uint128 liquidity,,) = _getPositionInfo();
+        uint128 liqudityToRemove = liquidity * withdrawLiquidityShare / 1e4;
+
+        // TODO add events
+        (uint256 amountWETH, uint256 amountOETHb) = positionManager.decreaseLiquidity(
+            INonfungiblePositionManager.DecreaseLiquidityParams({
+                tokenId: tokenId,
+                liquidity: liqudityToRemove,
+                /**
+                 * Both expected amounts can be 0 since we don't really care if any swaps
+                 * happen just before the liquidity removal.
+                 */
+                amount0Min: 0,
+                amount1Min: 0,
+                deadline: block.timestamp
+            })
+        );
+
+        positionManager.positions(tokenId);
+
+        positionManager.collect(
+            INonfungiblePositionManager.CollectParams({
+                tokenId: tokenId,
+                recipient: address(this),
+                amount0Max: type(uint128).max, // defaults to all tokens owed
+                amount1Max: type(uint128).max  // defaults to all tokens owed
+            })
+        );
+
+        // TODO can this go negative?
+        netValue -= amountWETH + amountOETHb;
+    }
+
+    /**
+     * @dev Check that the liquidity in the pool is withing the expected WETH to OETHb ratio
+     */
+    function _checkLiquidityWithinExpectedShare() internal {
+        if (tokenId == 0) {
+            return;
+        }
+        // TODO: calculate that the liquidiy shares are within the expected position
+        console.log("checking shares");
+
+        (uint256 amount0, uint256 amount1) = _getPositionPrincipal();
+        console.log(amount0);
+        console.log(amount1);
+
+        // calculate in 1 of 2 ways
+        // 1st way
+        // uint256 currentWethShare = amount0 * 1e4 / (amount0 + amount1)
+        // check that currentWethShare is within the poolWethShareVarianceAllowed of the
+        // configurable poolWethShare
+
+        // 2nd way
```

Comment:
I would lean towards a price check since this should be the least gas, and it's what's most important from a safety concern.

Reply:
talked about this one with @DanielVF async and we will stick with token shares for now